expo/expo-cli
# [config-plugins] Created unversioned react-native-maps plugin

# Why

- auto configure react-native-maps upon ejecting.

# How

- created an unversioned plugin for `react-native-maps` namespace which:
  - adds `pod 'react-native-google-maps', path: '../node_modules/react-native-maps'` to the podfile if `ios.config.googleMapsApiKey` is defined.
  - adds the following blocks to the AppDelegate.m when Google Maps is enabled:
```objc
// @generated begin react-native-maps-import - expo prebuild (DO NOT MODIFY) sync-f2f83125c99c0d74b42a2612947510c4e08c423a
#if __has_include(<GoogleMaps/GoogleMaps.h>)
#import <GoogleMaps/GoogleMaps.h>
#endif
// @generated end react-native-maps-import

...

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{

// @generated begin react-native-maps-init - expo prebuild (DO NOT MODIFY) sync-153e3eb24b7da99b846df4d93dd1d33da90e7c22
#if __has_include(<GoogleMaps/GoogleMaps.h>)
  [GMSServices provideAPIKey:@"API_KEY"];
#endif
// @generated end react-native-maps-init
```
These blocks are special because they can be removed or updated between prebuilds (plugin anti pattern). The `react-native-maps-init` block is matched against RN and Unimodules projects using the following (tested) regex `/(?:(self\.|_)(\w+)\s?=\s?\[\[UMModuleRegistryAdapter alloc\])|(?:RCTBridge\s?\*\s?(\w+)\s?=\s?\[\[RCTBridge alloc\])/g`

- The Android permissions `android.permission.ACCESS_COARSE_LOCATION` and `android.permission.ACCESS_FINE_LOCATION` are added
- The iOS permission `NSLocationWhenInUseUsageDescription` is added (no customization options)
- The AndroidManifest tags `com.google.android.geo.API_KEY` and `org.apache.http.legacy` are added, this was previously done directly but the plugin has now been moved into the unversioned `react-native-maps` plugin (meaning it can be disabled by a future plugin in the `react-native-maps` package).
- Finally, `GMSApiKey` is added to the Info.plist (not sure why though).



# Test Plan

- Extensive regex testing is done in the unit tests
- In a new project
- Install `yarn add react-native-maps`
- Sync `expod prebuild`
- Build `expod run:ios --bundler` -- this works as expected for Apple maps
- Define a value in `ios.config.googleMapsApiKey` 
- Sync `expod prebuild`
- Install `npx pod-install`
- Build `expod run:ios --bundler` -- this enables Google Maps on iOS
- `yarn remove react-native-maps`
- Sync `expod prebuild` -- this removes the related CocoaPods and AppDelegate code safely.
